feat(inputs.tacacs): Add tacacs plugin for simple tacacs auth response time monitoring

Hr0bar opened the pull request:
- [x] Updated associated README.md.
- [x] Wrote appropriate unit tests.
- [x] Pull request title or commits are in [conventional commit format](https://www.conventionalcommits.org/en/v1.0.0/#summary)

Opening PR after discussion in this Issue: https://github.com/influxdata/telegraf/issues/12573 and separating the tacacs plugin into a separate PR as discussed int https://github.com/influxdata/telegraf/pull/12736/

srebhan commented on plugins/inputs/tacacs/tacacs.go:
```diff
@@ -0,0 +1,164 @@
+	startTime := time.Now()
+	reply, session, err := client.SendAuthenStart(ctx, testAuthStart)
+	if err != nil {
+		t.Log.Warnf("error on new tacacs authentication start request to %s : %s", server, err)
+		fields["responsetime"] = time.Duration(t.ResponseTimeout).Seconds()
+		fields["responsetime_ms"] = time.Duration(t.ResponseTimeout).Milliseconds()
+		acc.AddFields("tacacs", fields, tags)
+		return nil
```
I think we should do the same as for the radius plugin. Remove one of the redundant `responsetime` fields and add some information to the metric indicating the "error-code"/"status".

srebhan commented on plugins/inputs/tacacs/tacacs.go:
```diff
@@ -0,0 +1,164 @@
+		return nil
+	}
+
+	// Send the first Continue request with the username, the reply should be AuthenStatusGetPass
+	reply, err = session.Continue(ctx, string(username))
+	if err != nil {
+		t.Log.Warnf("error on new tacacs authentication continue username request to %s : %s", server, err)
+		fields["responsetime"] = time.Duration(t.ResponseTimeout).Seconds()
+		fields["responsetime_ms"] = time.Duration(t.ResponseTimeout).Milliseconds()
+		acc.AddFields("tacacs", fields, tags)
+		return nil
+	}
+	if reply.Status != tacplus.AuthenStatusGetPass {
+		t.Log.Warnf("error on first tacacs continue username request to %s : Unexpected response code %d", server, reply.Status)
+		fields["responsetime"] = time.Duration(t.ResponseTimeout).Seconds()
+		fields["responsetime_ms"] = time.Duration(t.ResponseTimeout).Milliseconds()
+		acc.AddFields("tacacs", fields, tags)
+		return nil
+	}
+
+	reply, err = session.Continue(ctx, string(password))
+	if err != nil {
+		t.Log.Warnf("error on new tacacs authentication continue password request to %s : %s", server, err)
+		fields["responsetime"] = time.Duration(t.ResponseTimeout).Seconds()
+		fields["responsetime_ms"] = time.Duration(t.ResponseTimeout).Milliseconds()
+		acc.AddFields("tacacs", fields, tags)
+		return nil
+	}
```
Shouldn't those be _real_ errors for the plugin, i.e. return an error?

srebhan reviewed (changes requested):
Thanks for contributing this plugin @Hr0bar! I have some comments, mostly the same as for the radius one. Can you please also add a unit-test?!

Hr0bar commented:
Ill try to check on this likely over the coming weeks, likely most stuff that was done for the radius plugin will be applicable here as well.  Good thing the [tacacs GO package](https://github.com/nwaples/tacplus) seem to have some basic testing server implementation for tests and https://hub.docker.com/r/dchidell/docker-tacacs seem to be hopefully usable for integration tests.

powersj commented:
Hi @Hr0bar,

Any further progress on this PR recently?

Thanks!

Hr0bar commented:
Not yet, I shall have more time after April 22 (getting married, so not
much free time because of preparations)

On Tue, Apr 4, 2023 at 4:16 PM Joshua Powers ***@***.***>
wrote:

> Hi @Hr0bar <https://github.com/Hr0bar>,
>
> Any further progress on this PR recently?
>
> Thanks!
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/influxdata/telegraf/pull/12747#issuecomment-1496053590>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABTNGZQTLMJ62UBTPX75XEDW7QULLANCNFSM6AAAAAAVJNNOIE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

powersj commented:
> Not yet, I shall have more time after April 22

Thanks

> getting married, so not much free time because of preparations

Congrats! I know how busy that time can be :)

Hipska commented on plugins/inputs/tacacs/README.md:
````diff
@@ -0,0 +1,45 @@
+
+## Example Output
+
+```shell
+tacacs,source=debian-stretch-tacacs responsetime=0.011 1502489900000000000
````
This is not a shell script
```suggestion
```text
tacacs,source=debian-stretch-tacacs responsetime=0.011 1502489900000000000
```

Hipska commented on plugins/inputs/tacacs/README.md:
```diff
@@ -0,0 +1,45 @@
+- tacacs
+  - tags:
+    - source
+  - fields:
+    - responsetime (float)
```
In the code I also see `responsetime_ms`..

Hipska commented on plugins/inputs/tacacs/tacacs.go:
```diff
@@ -0,0 +1,164 @@
+		acc.AddFields("tacacs", fields, tags)
+		return nil
+	}
+	if reply.Status != tacplus.AuthenStatusGetPass {
+		t.Log.Warnf("error on first tacacs continue username request to %s : Unexpected response code %d", server, reply.Status)
+		fields["responsetime"] = time.Duration(t.ResponseTimeout).Seconds()
+		fields["responsetime_ms"] = time.Duration(t.ResponseTimeout).Milliseconds()
+		acc.AddFields("tacacs", fields, tags)
+		return nil
+	}
```
What about adding status in the returned metric?

Hipska replied:
> Tried using the status code returned, but it was very useless as the status depends on the current tacacs auth sequence. On initial start packet, return code X may mean something completely different than return code X for send username packet, then X may mean something else on send password packet etc. So without the secondary info at what stage the status code was returned, its useless. In the radius plugin, the status code is always from the same place - its just one request to do the auth, not multiple, so its nicely descriptive with exact string representation of the status. Here its not the case, you get an integer code that may mean something else for every auth stage.
> 
> Anyway, its meant to monitor SUCCESSFUL authentications response time, so decided to simply return an error in all other cases. Let me know if that makes sense ?

As far as I can see, the status code would only be one of these values?
https://github.com/nwaples/tacplus/blob/v0.0.3/packet.go#L62-L72

Hipska commented on plugins/inputs/tacacs/tacacs.go:
```diff
@@ -0,0 +1,164 @@
+	}
+	duration := time.Since(startTime)
+
+	if reply.Status != tacplus.AuthenStatusPass {
+		t.Log.Warnf("Got tacacs status code: %d", reply.Status)
+		fields["responsetime"] = time.Duration(t.ResponseTimeout).Seconds()
+		fields["responsetime_ms"] = time.Duration(t.ResponseTimeout).Milliseconds()
+	} else {
+		fields["responsetime"] = duration.Seconds()
+		fields["responsetime_ms"] = duration.Milliseconds()
+	}
```
Same here, status might be an interesting field?

Hipska commented on plugins/inputs/tacacs/tacacs.go:
```diff
@@ -0,0 +1,164 @@
+	if len(t.Servers) == 0 {
+		t.Servers = []string{"127.0.0.1"}
+	}
+	return nil
```
Maybe also check for other empty required fields? I think `RemAddr` and credentials need to be set?

Hipska replied:
> mandatory fields are uncommented in the sample conf. If not set, I think these should catch it:
> 
> ```
> 	username, err := t.Username.Get()
> 	if err != nil {
> 		return fmt.Errorf("getting username failed: %v", err)
> 	}
> ```

The problem is that these only get checked at runtime and result in an error each interval.

> The other non mandatory are now defaults: `return &Tacacs{RemAddr: "127.0.0.1", ResponseTimeout: config.Duration(time.Second * 5)}`

What if someone sets `request_ip = ""` or an invalid IP? A valid config still needs to be checked and alerted in `Init()`.

Hr0bar replied:
Im adding check for the remaining items (username, password, secret). Although this will fail on users that may want to use empty string for those fields (as empty string is zero bytes, which is considered same as not set by the Empty() check of the config.Secret type )

Hipska reviewed (changes requested):
A test is needed

Hr0bar commented:
BTW the "Git repo dirty. Please run "make docs" and push the updated README. Failing CI." in "ci/circleci: test-go-linux" is likely caused by me changing README.MD -> README.md filename in GIT some time ago in this PR. Ran make docs in the past after that, but still pops up

Hr0bar commented on plugins/inputs/tacacs/sample.conf:
```diff
@@ -0,0 +1,14 @@
+[[inputs.tacacs]]
```
Is this a new requirement since some months ago?  Thought its enough to have it in README.md (its also not present in the radius plugin in sample.conf)

Hipska commented on plugins/inputs/tacacs/tacacs.go:
```diff
@@ -0,0 +1,164 @@
+	}
+	defer config.ReleaseSecret(password)
+
+	// Create the tacacs client
+	client := &tacplus.Client{
+		Addr: server,
+		ConnConfig: tacplus.ConnConfig{
+			Secret: secret,
+		},
+	}
```
Second that! No need to recreate it every poll cycle.

Hr0bar replied:
Changed it to create it only once.

Hipska commented on plugins/inputs/tacacs/tacacs.go:
```diff
@@ -0,0 +1,147 @@
+	}
+	defer config.ReleaseSecret(password)
+
+	t.client.Addr = server
+	t.client.ConnConfig = tacplus.ConnConfig{
+		Secret: secret,
+	}
+
+	// Initialize the AuthStart data
+	testAuthStart := &tacplus.AuthenStart{
+		Action:        tacplus.AuthenActionLogin,
+		AuthenType:    tacplus.AuthenTypeASCII,
+		AuthenService: tacplus.AuthenServiceLogin,
+		PrivLvl:       1,
+		Port:          "heartbeat",
+		RemAddr:       t.RemAddr,
+	}
```
These can still be moved to Init, you can make an array of clients. Then the `pollServer` would get the `client tacplus.Client` or just the id..

Hipska commented on plugins/inputs/tacacs/tacacs.go:
```diff
@@ -0,0 +1,147 @@
+	defer session.Close()
+	// Check the returned status
+	if reply.Status != tacplus.AuthenStatusGetUser {
+		acc.AddFields("tacacs", fields, tags)
```
At this stage, you are sending a metric with no fields?

Hr0bar replied:
oh I missed this, I added that for testing the container tests, forgot to remove it afterwards :/ Only wanted to return errors and no metrics on hard error (hope thats the correct standard in telegraf)

Hipska replied:
Yes returning an error is the good way, no need to send an empty metric as that is not valid.

Hipska replied:
Please mark as resolved when done, I can't do this myself.

Hipska commented on plugins/inputs/tacacs/tacacs.go:
```diff
@@ -0,0 +1,147 @@
+	}
+	duration := time.Since(startTime)
+	if reply.Status != tacplus.AuthenStatusPass {
+		acc.AddFields("tacacs", fields, tags)
```
Same here, the metric has no fields..

srebhan commented:
Hey @Hr0bar,

regarding
> BTW the "Git repo dirty. Please run "make docs" and push the updated README. Failing CI." in "ci/circleci: test-go-linux" is likely caused by me changing README.MD -> README.md filename in GIT some time ago in this PR. Ran make docs in the past after that, but still pops up

running `make docs` should solve the issue. It's likely due to the `sample.conf` content doesn't match the README.md Configuration section content...

Hipska commented on plugins/inputs/tacacs/tacacs.go:
```diff
@@ -0,0 +1,150 @@
+		return fmt.Errorf("Invalid ip address provided for request_ip: %s", t.RemAddr)
+	}
+
+	for index, serverValue := range t.Servers {
+		t.clients = append(t.clients, tacplus.Client{})
```
I think you can already make the size of your `clients` array the same as the `Servers` array, so no need to append each loop.

Hr0bar commented on plugins/inputs/tacacs/tacacs.go:
```diff
@@ -0,0 +1,150 @@
+	tags := map[string]string{"source": client.Addr}
+	fields := make(map[string]interface{})
+
+	secret, err := t.Secret.Get()
+	if err != nil {
+		return fmt.Errorf("getting secret failed: %w", err)
+	}
+	defer config.ReleaseSecret(secret)
+
+	client.ConnConfig.Secret = secret
```
Did not study how exactly release secret works, but when I tried that the defer statement in Init() would clean the assigned value inside the client. Likely would have to to some sort of deep copy of the value perhaps to detach from the releasesecret functionality.

srebhan commented on plugins/inputs/tacacs/README.md:
```diff
@@ -0,0 +1,49 @@
+# Tacacs Input Plugin
+
+The Tacacs plugin collects successful tacacs authentication response times.
```
Is there an implementation/software we can or should reference?

Hr0bar replied:
Ill try mentioning at least aruba clearpass (which we use) and the one used in tests etc

srebhan commented on plugins/inputs/tacacs/tacacs.go:
```diff
@@ -0,0 +1,150 @@
+	Username        config.Secret   `toml:"username"`
+	Password        config.Secret   `toml:"password"`
+	Secret          config.Secret   `toml:"secret"`
+	RemAddr         string          `toml:"request_ip"`
```
Can we please use a more speaking name for the variable!?
```suggestion
	RequestAddr         string          `toml:"request_ip"`
```

Hr0bar replied:
Initially it was called like that because RemAddr naming is used in the underlying tacacs client library dependency, and thats the variable it sets. But I agree RemAddr is from the viewpoint of the tacacs server, for which the remote address is the telegraf agent. From telegraf agent viewpoint, this is our request address, so I agree that this should be changed.

When creating the client object, the old variable name will still appear in the code though:

```
tacplus.AuthenStart{
		Action:        tacplus.AuthenActionLogin,
		AuthenType:    tacplus.AuthenTypeASCII,
		AuthenService: tacplus.AuthenServiceLogin,
		PrivLvl:       1,
		Port:          "heartbeat",
		RemAddr:       t.RequestAddr,
	}
```

srebhan commented on plugins/inputs/tacacs/tacacs.go:
```diff
@@ -0,0 +1,150 @@
+	ResponseTimeout config.Duration `toml:"response_timeout"`
+	Log             telegraf.Logger `toml:"-"`
+	clients         []tacplus.Client
+	testAuthStart   tacplus.AuthenStart
```
Why does it have a `test` prefix? Can we simply name it `authStart` or `auth` or similar?

Hr0bar replied:
I dont think theres a reason other than it was written some years ago and it stayed the same. Changing it to authStart is sensible.

srebhan commented on plugins/inputs/tacacs/tacacs.go:
```diff
@@ -0,0 +1,150 @@
+	if len(t.Servers) == 0 {
+		t.Servers = []string{"127.0.0.1:49"}
+	}
+	if net.ParseIP(t.RemAddr) == nil {
+		return fmt.Errorf("Invalid ip address provided for request_ip: %s", t.RemAddr)
+	}
```
Your `sample.conf` says the default is `127.0.0.1` and I think that should be set here (instead of in `init()`) if empty. Furthermore, errors should start with lowercase letters in Golang...

Hr0bar replied:
Sure I can move the default to Init() instead of init() if thats the usual way in telegraf.

But this error check is for when user enters invalid IP address, not when he leaves it empty. So Ill set default just above it.

srebhan commented on plugins/inputs/tacacs/tacacs.go:
```diff
@@ -0,0 +1,150 @@
+		return fmt.Errorf("Invalid ip address provided for request_ip: %s", t.RemAddr)
+	}
+
+	for index, serverValue := range t.Servers {
+		t.clients = append(t.clients, tacplus.Client{})
+		t.clients[index].Addr = serverValue
+		t.clients[index].ConnConfig = tacplus.ConnConfig{}
+	}
```
Maybe
```suggestion
	t.clients = make([]tacplug.Client, 0, len(t.Servers)) 
	for _, server := range t.Servers {
		t.clients = append(t.clients, tacplus.Client{
			Addr: server,
			ConnConfig: tacplus.ConnConfig{},
		})
	}
```

srebhan commented on plugins/inputs/tacacs/tacacs.go:
```diff
@@ -0,0 +1,150 @@
+		return fmt.Errorf("error on new tacacs authentication start request to %s : %w", client.Addr, err)
+	}
+	defer session.Close()
+	// Check the returned status
+	if reply.Status != tacplus.AuthenStatusGetUser {
+		return fmt.Errorf("error on new tacacs authentication start request to %s : Unexpected response code %d", client.Addr, reply.Status)
+	}
```
I don't remember what we did on radius, but it might make sense to have this in the metric to make it discoverable to monitoring....

Hr0bar replied:
in radius we set responsetime to maximum timeout value if its not success and use response code in tag, Ill do that, although its harder to debug for users based on that info due to the multi-stage auth rather than single stage as in radius. (He wont have info from which stage the response code is - that is currently provided by the hard error message). Optionally we could do custom status from this set: ["success", "error on new tacacs authentication start request to %s : Unexpected response code %d", "error on first tacacs authentication continue username request to %s : Unexpected response code %d", "error on second tacacs authentication continue password request to %s : Unexpected response code %d"] which would provide more immediate info. Let me know what you think. For now commiting it with the raw values.

srebhan commented on plugins/inputs/tacacs/tacacs_test.go:
```diff
@@ -0,0 +1,244 @@
+	"testing"
+	"time"
+
+	"github.com/influxdata/telegraf/config"
+	"github.com/influxdata/telegraf/testutil"
+	"github.com/nwaples/tacplus"
+	"github.com/stretchr/testify/require"
+	"github.com/testcontainers/testcontainers-go/wait"
```
```suggestion
	"github.com/nwaples/tacplus"
	"github.com/stretchr/testify/require"
	"github.com/testcontainers/testcontainers-go/wait"

	"github.com/influxdata/telegraf/config"
	"github.com/influxdata/telegraf/testutil"
```

Hr0bar commented:
Thanks, Ill check these comments when Ill be back from travelling ~mid
June. Or feel free to adjust it until then.

On Wed, May 17, 2023 at 3:44 PM Sven Rebhan ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
>
> @Hr0bar <https://github.com/Hr0bar> thanks for the update. Just a few
> minor comments...
> ------------------------------
>
> In plugins/inputs/tacacs/README.md
> <https://github.com/influxdata/telegraf/pull/12747#discussion_r1196529570>
> :
>
> > @@ -0,0 +1,49 @@
> +# Tacacs Input Plugin
> +
> +The Tacacs plugin collects successful tacacs authentication response times.
>
> Is there an implementation/software we can or should reference?
> ------------------------------
>
> In plugins/inputs/tacacs/tacacs.go
> <https://github.com/influxdata/telegraf/pull/12747#discussion_r1196532116>
> :
>
> > +	"sync"
> +	"time"
> +
> +	"github.com/nwaples/tacplus"
> +
> +	"github.com/influxdata/telegraf"
> +	"github.com/influxdata/telegraf/config"
> +	"github.com/influxdata/telegraf/plugins/inputs"
> +)
> +
> +type Tacacs struct {
> +	Servers         []string        `toml:"servers"`
> +	Username        config.Secret   `toml:"username"`
> +	Password        config.Secret   `toml:"password"`
> +	Secret          config.Secret   `toml:"secret"`
> +	RemAddr         string          `toml:"request_ip"`
>
> Can we please use a more speaking name for the variable!?
> ⬇️ Suggested change
>
> -	RemAddr         string          `toml:"request_ip"`
> +	RequestAddr         string          `toml:"request_ip"`
>
> ------------------------------
>
> In plugins/inputs/tacacs/tacacs.go
> <https://github.com/influxdata/telegraf/pull/12747#discussion_r1196534235>
> :
>
> > +
> +	"github.com/influxdata/telegraf"
> +	"github.com/influxdata/telegraf/config"
> +	"github.com/influxdata/telegraf/plugins/inputs"
> +)
> +
> +type Tacacs struct {
> +	Servers         []string        `toml:"servers"`
> +	Username        config.Secret   `toml:"username"`
> +	Password        config.Secret   `toml:"password"`
> +	Secret          config.Secret   `toml:"secret"`
> +	RemAddr         string          `toml:"request_ip"`
> +	ResponseTimeout config.Duration `toml:"response_timeout"`
> +	Log             telegraf.Logger `toml:"-"`
> +	clients         []tacplus.Client
> +	testAuthStart   tacplus.AuthenStart
>
> Why does it have a test prefix? Can we simply name it authStart or auth
> or similar?
> ------------------------------
>
> In plugins/inputs/tacacs/tacacs.go
> <https://github.com/influxdata/telegraf/pull/12747#discussion_r1196537058>
> :
>
> > +	if net.ParseIP(t.RemAddr) == nil {
> +		return fmt.Errorf("Invalid ip address provided for request_ip: %s", t.RemAddr)
> +	}
>
> Your sample.conf says the default is 127.0.0.1 and I think that should be
> set here (instead of in init()) if empty. Furthermore, errors should
> start with lowercase letters in Golang...
> ------------------------------
>
> In plugins/inputs/tacacs/tacacs.go
> <https://github.com/influxdata/telegraf/pull/12747#discussion_r1196540831>
> :
>
> > +	for index, serverValue := range t.Servers {
> +		t.clients = append(t.clients, tacplus.Client{})
> +		t.clients[index].Addr = serverValue
> +		t.clients[index].ConnConfig = tacplus.ConnConfig{}
> +	}
>
> Maybe
> ⬇️ Suggested change
>
> -	for index, serverValue := range t.Servers {
> -		t.clients = append(t.clients, tacplus.Client{})
> -		t.clients[index].Addr = serverValue
> -		t.clients[index].ConnConfig = tacplus.ConnConfig{}
> -	}
> +	t.clients = make([]tacplug.Client, 0, len(t.Servers))
> +	for _, server := range t.Servers {
> +		t.clients = append(t.clients, tacplus.Client{
> +			Addr: server,
> +			ConnConfig: tacplus.ConnConfig{},
> +		})
> +	}
>
> ------------------------------
>
> In plugins/inputs/tacacs/tacacs.go
> <https://github.com/influxdata/telegraf/pull/12747#discussion_r1196545689>
> :
>
> > +	// Check the returned status
> +	if reply.Status != tacplus.AuthenStatusGetUser {
> +		return fmt.Errorf("error on new tacacs authentication start request to %s : Unexpected response code %d", client.Addr, reply.Status)
> +	}
>
> I don't remember what we did on radius, but it might make sense to have
> this in the metric to make it discoverable to monitoring....
> ------------------------------
>
> In plugins/inputs/tacacs/tacacs_test.go
> <https://github.com/influxdata/telegraf/pull/12747#discussion_r1196546692>
> :
>
> > +	"github.com/influxdata/telegraf/config"
> +	"github.com/influxdata/telegraf/testutil"
> +	"github.com/nwaples/tacplus"
> +	"github.com/stretchr/testify/require"
> +	"github.com/testcontainers/testcontainers-go/wait"
>
> ⬇️ Suggested change
>
> -	"github.com/influxdata/telegraf/config"
> -	"github.com/influxdata/telegraf/testutil"
> -	"github.com/nwaples/tacplus"
> -	"github.com/stretchr/testify/require"
> -	"github.com/testcontainers/testcontainers-go/wait"
> +	"github.com/nwaples/tacplus"
> +	"github.com/stretchr/testify/require"
> +	"github.com/testcontainers/testcontainers-go/wait"
> +
> +	"github.com/influxdata/telegraf/config"
> +	"github.com/influxdata/telegraf/testutil"
>
> ------------------------------
>
> In plugins/inputs/tacacs/tacacs.go
> <https://github.com/influxdata/telegraf/pull/12747#discussion_r1196548571>
> :
>
> > +	ResponseTimeout config.Duration `toml:"response_timeout"`
> +	Log             telegraf.Logger `toml:"-"`
> +}
> +
> +//go:embed sample.conf
> +var sampleConfig string
> +
> +func (t *Tacacs) SampleConfig() string {
> +	return sampleConfig
> +}
> +
> +func (t *Tacacs) Init() error {
> +	if len(t.Servers) == 0 {
> +		t.Servers = []string{"127.0.0.1"}
> +	}
> +	return nil
>
> I do agree, it might be worth checking the options in Init() to fail
> early and loudly!
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/influxdata/telegraf/pull/12747#pullrequestreview-1430713678>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABTNGZU6KUCQPZGW4YF7TYTXGTI5JANCNFSM6AAAAAAVJNNOIE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

srebhan commented:
Ok. It's a pitty as we will likely miss v1.27.0 which is planned for June 12th... :-(

srebhan commented:
@Hr0bar any chance you work on this PR?

Hr0bar commented:
I added all the changes, BUT the one regarding returning the return status in case of failures likely needs more thought. I added the response_code tag, and returning timeout value for response time, same as in the other plugin. That is all OK, but it works for wrong password, wrong username etc, but it doesnt cover for example wrong tacacs secret. In that case client.SendAuthenStart returns hard error, and no response code is returned. I think it should be consistent for the user, so as I commented on that suggestion already, what about creating the response codes completely custom and human readable for telegraf purposes? Instead of the hard errors there were previously.

Hr0bar commented:
So either:

1. restore parts of code https://github.com/influxdata/telegraf/pull/12747/commits/4387d9f7411e067bee2e07be8d8ae702745ce1c8 from before today for cases where wrong password etc is used, that return hard error on everything that is not success
2. keep the current weird mix where we return the response code in tag, but it doesnt apply to wrong secret and possibly other cases, because in that case tacacs servers dont return any response and there is hard error on connection instead.
3. Implement custom response codes, that provide all info to the user (including real response codes if provided by tacacs server), that cover also wrong tacacs secret and other cases. Use timeout time for response time in case of those failures (same as in the other plugin)

srebhan commented:
> 1. restore parts of code [4387d9f](https://github.com/influxdata/telegraf/commit/4387d9f7411e067bee2e07be8d8ae702745ce1c8) from before today for cases where wrong password etc is used, that return hard error on everything that is not success
> 2. keep the current weird mix where we return the response code in tag, but it doesnt apply to wrong secret and possibly other cases, because in that case tacacs servers dont return any response and there is hard error on connection instead.
> 3. Implement custom response codes, that provide all info to the user (including real response codes if provided by tacacs server), that cover also wrong tacacs secret and other cases. Use timeout time for response time in case of those failures (same as in the other plugin)

srebhan commented on plugins/inputs/tacacs/sample.conf:
```diff
@@ -0,0 +1,15 @@
+  secret = "mysecret"
+
+  ## Maximum time to receive response.
+  # response_timeout = "5s"
```
Please add a newline at the end of the file.

Hr0bar replied:
added newline, added condition for timeout

srebhan reviewed (commented):
Thanks @Hr0bar for the update! The code looks good to me and to me the current response-code strategy also seems good. IMO we error out if the _sending_ reports an error. If the sending succeeds but the server reports some unexpected state we pass that information on to the user...

The only comment I have is that currently we do _not_ report the (context/connection) timeout to the user. I think you should make an exception there when checking the error codes and also generate a metric for this case. This way an operator can see that Telegraf cannot reach the server and check what is going on...

Hipska commented on plugins/inputs/tacacs/tacacs_test.go:
```diff
@@ -0,0 +1,255 @@
+			}
+
+			if tt.name == "wrong_pw" {
+				require.NoError(t, plugin.Init())
+				require.NoError(t, plugin.Gather(&acc))
```
Double Init and Gather?

Hr0bar replied:
Good catch, removed it.

Hipska commented on plugins/inputs/tacacs/tacacs_test.go:
```diff
@@ -0,0 +1,255 @@
+			require.NoError(t, plugin.Gather(&acc))
+
+			if tt.expectSuccess {
+				if len(acc.Errors) > 0 {
+					t.Errorf("error occured in test where should be none, error was: %s", acc.Errors[0].Error())
+				}
```
```suggestion
				require.Lenf(t, acc.Errors, 0, "error occurred in test where should be none, error was: %s", acc.Errors[0].Error())
```

Hr0bar replied:
changed it to use require.Lenf as suggested

Hr0bar replied:
Reverted it, seems t.Errorf evaluates/constructs the string even if no error occurs (I would expect it to get constructed only when needed), causing index out of range

Hipska replied:
Strange, then why does it work in the `if tt.name == "wrong_pw" {` section?

Hr0bar replied:
There acc.Errors[0].Error() is not sent as a parameter (when no error exists, acc.Errors is empty, so no index 0 is present)

Hipska replied:
Right, then it should be `acc.FirstError()` or simply this:
```suggestion
				require.NoError(t, acc.FirstError())
```

Hr0bar replied:
Thats a good helper function (FirstError), changed it, will appear in next push (hopefully wont be many left)

srebhan commented on plugins/inputs/tacacs/tacacs_test.go:
```diff
@@ -0,0 +1,255 @@
+			Mux:    true,
+		},
+	}
+	l, err := net.Listen("tcp", "localhost:1049") // Use port 1049 instead of 49, so we are above 1023 for unix tests
+	require.NoError(t, err, "local net listen failed to start listening on port 1049")
```
Yeah you can use `localhost:0` and will get a random free port.

Hr0bar replied:
modified it to use localhost:0

Hr0bar commented on plugins/inputs/tacacs/tacacs_test.go:
```diff
@@ -0,0 +1,255 @@
+		Log:             testutil.Logger{},
+	}
+	var acc testutil.Accumulator
+	require.NoError(t, plugin.Init())
```
added some Init tests (tried also default timeout but thats set in init() instead of Init() so its not called in tests, would need more effort to implement test for that)

Hipska replied:
I think its best to have that in a separate test like `TestInit`..

Hr0bar replied:
yea makes it more readable

Hr0bar commented on plugins/inputs/tacacs/tacacs_test.go:
```diff
@@ -0,0 +1,255 @@
+		ResponseTimeout: config.Duration(time.Second * 5),
+		Log:             testutil.Logger{},
+	}
+	var acc2 testutil.Accumulator
```
using array of tests now

Hr0bar commented on plugins/inputs/tacacs/tacacs.go:
```diff
@@ -0,0 +1,169 @@
+	}
+	defer session.Close()
+	if reply.Status != tacplus.AuthenStatusGetUser {
+		fields["responsetime_ms"] = time.Duration(t.ResponseTimeout).Milliseconds()
```
Honestly I have hard time deciding whats best. For our use case how we use it - we are only interested in successful auth response times, if its not successful, the response time is irrelevant to us and warrants an alert creation (whether by bad response code, or by longer than allowed response time). But agree perhaps some users might find it useful for different use cases. Now its possible to check response_code==1 for successful only, so we can live with anything here.

Wondering if context.DeadlineExceeded err can be hit with timeout set to 0, because then we would be filling the response time with 0

Hipska replied:
Yeah, I do think its expected to have a real response time as it might indicate something why authentication cannot continue maybe?

Hr0bar replied:
` I'm not sure if the status code should be a tag. Looks more like it should be a field to me. Or is it to be expected it to always be the same for a given configuration?`  

It can change if the remote server goes down, users password changes on remote server etc. If all stays the same on both client and server sides, should be always the same value.

Hipska replied:
So it is a monitoring status and thus should be a field..

Hr0bar replied:
Im adding it as a field then, also outputting real response time if there is ANY response. Added notes to README so its clear whats happening.

Im also a bit tempted to replace the hard errors thrown by the client after SendAuthenStart with a custom response_code of "error" similar to "timeout", simply because they can happen often due to similar reasons to timeout (issues with network, even wrong SECRET causes `return fmt.Errorf("error on new tacacs authentication start request to %s : %w", client.Addr, err)` while wrong password and wrong username cause regular response.
So instead of current:
| Field value | Meaning                   | From          | responsetime_ms
| ----------- | ------------------------- | ------------- | ---------------
| 1           | AuthenStatusPass          | tacacs server | real value
| 2           | AuthenStatusFail          | tacacs server | real value
| 3           | AuthenStatusGetData       | tacacs server | real value
| 4           | AuthenStatusGetUser       | tacacs server | real value
| 5           | AuthenStatusGetPass       | tacacs server | real value
| 6           | AuthenStatusRestart       | tacacs server | real value
| 7           | AuthenStatusError         | tacacs server | real value
| 33          | AuthenStatusFollow        | tacacs server | real value
| timeout     | timeout                   | telegraf      | eq. to response_timeout

we would also have
| Field value | Meaning                   | From          | responsetime_ms
| ----------- | ------------------------- | ------------- | ---------------
| error | error | telegraf      | eq. to response_timeout

Not sure where we would output the actual error details, whether into warn telegraf log, or into some additional field that is either ok or the error message or similar

Hipska replied:
Ouch, that would mean the type suddenly will change from integer to string. A lot of outputs will not be able to handle that..

I would suggest to use `-1` instead.

Hr0bar replied:
Added const reverse translation function (since underlying dependency does not offer it), and returning human readable code string representation now.

Hipska commented on plugins/inputs/tacacs/tacacs_test.go:
```diff
@@ -0,0 +1,353 @@
+	}
+}
+
+func TestTacacsInit(t *testing.T) {
```
I would add this as the first test, so other tests don't fail on `plugin.Init()`

Hr0bar replied:
moved it to top

Hipska commented on plugins/inputs/tacacs/tacacs_test.go:
```diff
@@ -0,0 +1,353 @@
+func TestTacacsInit(t *testing.T) {
+	var testset = []struct {
+		name               string
+		testingTimeout     config.Duration
+		serverToTest       []string
+		usedUsername       config.Secret
+		usedPassword       config.Secret
+		usedSecret         config.Secret
+		requestAddr        string
+		expectNoInitErrors bool
```
What would you think of this?
```suggestion
		plugin      Tacacs
		errContains string
```
(+ next suggestion)

Hipska commented on plugins/inputs/tacacs/tacacs_test.go:
```diff
@@ -0,0 +1,353 @@
+
+	for _, tt := range testset {
+		t.Run(tt.name, func(t *testing.T) {
+			plugin := &Tacacs{
+				ResponseTimeout: tt.testingTimeout,
+				Servers:         tt.serverToTest,
+				Username:        tt.usedUsername,
+				Password:        tt.usedPassword,
+				Secret:          tt.usedSecret,
+				RequestAddr:     tt.requestAddr,
+				Log:             testutil.Logger{},
+			}
+
+			if tt.expectNoInitErrors {
+				require.NoError(t, plugin.Init())
+			} else {
+				initErr := plugin.Init()
+				require.Error(t, initErr)
+				if tt.name == "empty_creds" {
+					require.ErrorContains(t, initErr, "empty credentials were provided (username, password or secret)")
+				}
+				if tt.name == "wrong_reqaddress" {
+					require.ErrorContains(t, initErr, "invalid ip address provided for request_ip")
+				}
+			}
+
+			if tt.name == "no_reqaddress" {
+				require.Equal(t, "127.0.0.1", plugin.RequestAddr)
+			}
```
```suggestion
			t.plugin.Log = testutil.Logger{}
			
			err := t.plugin.Init()
			
			if tt.errContains == "" {
				require.NoError(t, err)
			} else {
				require.ErrorContains(t, err, t.errorContains)
			}
```

As it doesn't seem a best practice to have these if-cases for tests, better to have it all defined in the testset.

Hr0bar replied:
Ill have a look over the coming days, also at RequireMetricEqual instead of multiple separate statements

Hr0bar replied:
Removed all if-cases based on test name, now acts on testset definition

Hipska commented on plugins/inputs/tacacs/tacacs_test.go:
```diff
@@ -0,0 +1,353 @@
+				require.Len(t, acc.Errors, 0)
+			}
+
+			if tt.name == "wrong_pw" {
```
Actually, same remark can go here..

Hipska commented on plugins/inputs/tacacs/README.md:
```diff
@@ -46,20 +46,20 @@ See the [CONFIGURATION.md][CONFIGURATION.md] for more details.
+| AuthenStatusRestart  | 6 (0x6)      | tacacs server | real value
+| AuthenStatusError    | 7 (0x7)      | tacacs server | real value
+| AuthenStatusFollow   | 33 (0x21)    | tacacs server | real value
+| timeout              | timeout      | telegraf      | eq. to response_timeout
```
Maybe capital to be consistent with the other statuses?
```suggestion
| Timeout              | timeout      | telegraf      | eq. to response_timeout
```

Hipska commented on plugins/inputs/tacacs/README.md:
```diff
@@ -0,0 +1,75 @@
+  - tags:
+    - source
+  - fields:
+    - response_code (string, [see below](#field-response_code)))
```
Doesn't feel like a code anymore. Maybe rename to `response_status`?

Hipska commented on plugins/inputs/tacacs/README.md:
```diff
@@ -0,0 +1,75 @@
+
+### field `responsetime_ms`
+
+The field responsetime_ms is responsetime of the tacacs server
```
```suggestion
The field responsetime_ms is response time of the tacacs server
```

Hr0bar commented:
Let me know, if the tests are readable/by telegraf standards enough, or if additional refactoring is needed

Hipska commented on plugins/inputs/tacacs/tacacs.go:
```diff
@@ -149,14 +149,14 @@ func (t *Tacacs) pollServer(acc telegraf.Accumulator, client *tacplus.Client) er
 		}
 		fields["responsetime_ms"] = time.Duration(t.ResponseTimeout).Milliseconds()
-		fields["response_code"] = "timeout"
+		fields["response_status"] = "timeout"
```
```suggestion
		fields["response_status"] = "Timeout"
```

Hipska commented on plugins/inputs/tacacs/tacacs.go:
```diff
@@ -167,13 +167,13 @@ func (t *Tacacs) pollServer(acc telegraf.Accumulator, client *tacplus.Client) er
 		}
 		fields["responsetime_ms"] = time.Duration(t.ResponseTimeout).Milliseconds()
-		fields["response_code"] = "timeout"
+		fields["response_status"] = "timeout"
```
```suggestion
		fields["response_status"] = "Timeout"
```

Hipska commented on plugins/inputs/tacacs/tacacs_test.go:
```diff
@@ -156,54 +150,55 @@ func TestTacacsLocal(t *testing.T) {
-			expectNoGatherErrors: false,
+			name:           "unreachable",
+			testingTimeout: config.Duration(time.Nanosecond * 1000),
+			serverToTest:   []string{"unreachable.hostname.com:404"},
+			usedUsername:   config.NewSecret([]byte(`testusername`)),
+			usedPassword:   config.NewSecret([]byte(`testpassword`)),
+			usedSecret:     config.NewSecret([]byte(`testsecret`)),
+			requestAddr:    "127.0.0.1",
+			errContains:    "error on new tacacs authentication start request to unreachable.hostname.com:404 : dial tcp",
```
hostname.com is actually in use, so use one of the reserved names. Example:
```suggestion
			serverToTest:   []string{"unreachable.test:49"},
			usedUsername:   config.NewSecret([]byte(`testusername`)),
			usedPassword:   config.NewSecret([]byte(`testpassword`)),
			usedSecret:     config.NewSecret([]byte(`testsecret`)),
			requestAddr:    "127.0.0.1",
			errContains:    "error on new tacacs authentication start request to unreachable.test:49 : dial tcp",
```

Hipska reviewed (changes requested):
Yup, looks good. Just a few minor things..

srebhan reviewed (approved):
Looks good to me. Thanks for contributing this plugin @Hr0bar!

powersj reviewed (commented):
One question inline to help me understand the reported value for response time.

Hr0bar commented on plugins/inputs/tacacs/tacacs.go:
```diff
@@ -0,0 +1,208 @@
+		if !errors.Is(err, context.DeadlineExceeded) {
+			return fmt.Errorf("error on tacacs authentication continue username request to %s : %w", client.Addr, err)
+		}
+		fields["responsetime_ms"] = time.Duration(t.ResponseTimeout).Milliseconds()
```
I think that is a valid point, especially for users who would set the timeout to zero. In other cases, it will be close to the configured timeout value anyway. With zero timeout configured it can theoretically also timeout from other sources than the configured context deadline (from some network functions called down the line). Maybe in addition to context.DealineExceeded error we should also allow errors.Is(err, os.ErrDeadlineExceeded) for these cases.

Hr0bar replied:
While the used tacacs package seems to be correctly setting the timeout to the deeper used network calls, I could not deterministically say that it will always return only context.DeadlineExceeded error on timeout. 

It Seemd os.IsTimeout() is would be better here, as it returns True for both context.DeadlineExceeded and os.ErrDeadlineExceeded and possibly other timeout time errors, but after testing it it was too lenient/broad and it was returning True even for immediate dial tcp or dns errors etc so will keep 
`if !errors.Is(err, context.DeadlineExceeded) && !errors.Is(err, os.ErrDeadlineExceeded)`

The timeout time returned is now real value.

powersj reviewed (approved):
Thank you for the effort and work on this!
